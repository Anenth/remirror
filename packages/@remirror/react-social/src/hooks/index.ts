export * from './use-social';
export * from './use-social-emoji';
export * from './use-social-mention';
