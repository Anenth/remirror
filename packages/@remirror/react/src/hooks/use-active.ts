import type { AnyExtension, RemirrorManager } from '@remirror/core';

import { ReactFrameworkOutput } from '../react-types';
import { useRemirrorContext } from './use-remirror-context';

type UseActiveReturn<Extension extends AnyExtension> = ReactFrameworkOutput<
  RemirrorManager<Extension>['~E']
>['active'];

/**
 * This is a shorthand method for retrieving the active available in the
 * editor.
 *
 * ```ts
 * import { useActive } from 'remirror/react';
 * ```
 *
 * This hooks updates the local component on each state update for the editor,
 * so it can be quite expensive.
 *
 * @param autoUpdate - Set to false to prevent automatic re-rendering on every
 * state update.
 */
export function useActive<Extension extends AnyExtension = Remirror.Extensions>(
  autoUpdate = true,
): UseActiveReturn<Extension> {
  return useRemirrorContext<Extension>({ autoUpdate }).active;
}
