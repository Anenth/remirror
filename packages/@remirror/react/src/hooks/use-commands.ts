import { AnyExtension, ChainedFromExtensions, CommandsFromExtensions } from '@remirror/core';

import { useRemirrorContext } from './use-remirror-context';

/**
 * A core hook which provides the commands for usage in your editor.
 *
 * ```tsx
 * import { useCommands } from 'remirror/react';
 *
 * const EditorButton = () => {
 *   const commands = useCommands();
 *
 *   return (
 *     <>
 *       <button onClick={() => commands.toggleBold()}>
 *         Click me!
 *       </button>
 *     </>
 *   );
 * }
 * ````
 */
export function useCommands<
  Extension extends AnyExtension = Remirror.Extensions
>(): CommandsFromExtensions<Extension> {
  return useRemirrorContext<Extension>().commands;
}

/**
 * A core hook which provides the chainable commands for usage in your editor.
 *
 * ```tsx
 * import { useCommandsChain } from 'remirror/react';
 *
 * const EditorButton = () => {
 *   const chain = useCommandsChain();
 *
 *   return (
 *     <>
 *       <button onClick={() => chain.toggleBold().toggleItalic().run()}>
 *         Chain me!
 *       </button>
 *     </>
 *   );
 * }
 * ````
 */
export function useCommandsChain<
  Extension extends AnyExtension = Remirror.Extensions
>(): ChainedFromExtensions<Extension> {
  return useRemirrorContext<Extension>().chain;
}
