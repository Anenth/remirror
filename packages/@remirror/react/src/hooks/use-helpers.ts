import type { AnyExtension, RemirrorManager } from '@remirror/core';

import { ReactFrameworkOutput } from '../react-types';
import { useRemirrorContext } from './use-remirror-context';

type UseHelpersReturn<ExtensionUnion extends AnyExtension> = ReactFrameworkOutput<
  RemirrorManager<ExtensionUnion>['~E']
>['helpers'];

/**
 * A core hook which provides the helpers for usage in your editor.
 *
 * ```tsx
 * import { useHelpers } from 'remirror/react';
 *
 * const EditorButton = () => {
 *   const helpers = useHelpers();
 *
 *   return (
 *     <>
 *       <button onClick={() => Helpers.toggleBold()}>
 *         Click me!
 *       </button>
 *       <button onClick={() => Helpers.chain.toggleBold().toggleItalic().run()}>
 *         Chain me!
 *       </button>
 *     </>
 *   );
 * }
 * ````
 *
 * Passing `true` as the first argument will ensure that the component this hook
 * is placed inside will rerender on every update.
 */
export function useHelpers<ExtensionUnion extends AnyExtension = Remirror.AllExtensionUnion>(
  update = false,
): UseHelpersReturn<ExtensionUnion> {
  return useRemirrorContext<ExtensionUnion>(update ? { autoUpdate: true } : undefined).helpers;
}
