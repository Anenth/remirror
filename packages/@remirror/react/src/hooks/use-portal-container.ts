/**
 * @module
 *
 * A hook which provides access to the portal container for rendering react
 * component directly within the editor dom.
 */

import { PortalContainer } from '@remirror/extension-react-component';

import { useRemirrorContext } from './use-remirror-context';

/**
 * Retrieve the portal container which can be used to render
 */
export function usePortalContainer(): PortalContainer {
  return useRemirrorContext().portalContainer;
}
