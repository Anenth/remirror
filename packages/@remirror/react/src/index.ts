export type { UseExtensionCallback, UseRemirrorProps, UseRemirrorReturn } from './hooks';
export {
  RemirrorContext,
  useActive,
  useCommands,
  useEditorDomRef,
  useEditorState,
  useEffectWithWarning,
  useExtension,
  useForceUpdate,
  useHelpers,
  useI18n,
  useManager,
  usePortalContainer,
  useRemirror,
  useRemirrorContext,
} from './hooks';
export { createReactManager } from './react-helpers';
export type { RemirrorProps } from './react-remirror';
export { EditorComponent, Remirror } from './react-remirror';
export type {
  CreateReactManagerOptions,
  DefaultReactExtensionUnion,
  GetRootPropsConfig,
  ReactExtensionUnion,
  ReactFrameworkOutput,
  RefKeyRootProps,
  RefProps,
  UseRemirrorContextType,
} from './react-types';
export * from './renderers';
export { createEditorView } from './ssr';
