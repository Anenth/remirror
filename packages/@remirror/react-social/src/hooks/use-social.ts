import { useEffect, useRef } from 'react';

import { AnyCombinedUnion, isRemirrorManager, RemirrorManager } from '@remirror/core';
import { useRemirror, UseRemirrorType } from '@remirror/react';

import { CreateSocialManagerOptions, SocialCombinedUnion } from '../social-types';
import { createSocialManager } from '../social-utils';

/**
 * A wrapper around the `createSocialManager` function for creating a manager
 * within a react context.
 *
 * @remarks
 *
 * This can be useful if you want the manager to respond to take in
 * configuration from props.
 *
 * Please note that at the moment the manager returned is a ref. This means that
 * it won't be updated for the lifetime of the component. It takes an initial
 * configuration and then ignores everything else.
 */
export function useSocialManager<Combined extends AnyCombinedUnion>(
  managerOrCombined: readonly Combined[] | RemirrorManager<Combined | SocialCombinedUnion>,
  options: CreateSocialManagerOptions = {},
): RemirrorManager<SocialCombinedUnion | Combined> {
  const manager = useRef(
    isRemirrorManager<Combined | SocialCombinedUnion>(managerOrCombined)
      ? managerOrCombined
      : createSocialManager(managerOrCombined, options),
  ).current;

  useEffect(() => {
    return () => {
      manager.destroy();
    };
  }, [manager]);

  return manager;
}

/**
 * A properly typed alternative `useRemirror` hook for the social editor.
 */
export const useSocialRemirror = useRemirror as UseRemirrorType<SocialCombinedUnion>;
