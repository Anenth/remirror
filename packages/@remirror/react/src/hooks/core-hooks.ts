import isEqual from 'fast-deep-equal/react';
import { DependencyList, useEffect, useLayoutEffect, useRef, useState } from 'react';
import useEffectOnce from 'react-use/lib/useEffectOnce';
import useSetState from 'react-use/lib/useSetState';
import useUpdateEffect from 'react-use/lib/useUpdateEffect';
import ResizeObserver from 'resize-observer-polyfill';

/**
 * Preserves the previous version of a provided value.
 *
 * ```tsx
 * const [isOpen, setOpen] = useState<boolean>(false)
 * const previous = usePrevious(isOpen)
 *
 * return <span onClick={() => setOpen(!isOpen)}>{isOpen && previous === isOpen ? 'Stable' : 'Unstable' }</span>
 * ```
 */
export function usePrevious<Value>(value: Value) {
  const ref = useRef<Value>();
  useEffect(() => void (ref.current = value), [value]);
  return ref.current;
}

export interface DOMRectReadOnlyLike {
  readonly x: number;
  readonly y: number;
  readonly width: number;
  readonly height: number;
  readonly top: number;
  readonly right: number;
  readonly bottom: number;
  readonly left: number;
}

const defaultBounds = { x: 0, y: 0, width: 0, height: 0, top: 0, right: 0, bottom: 0, left: 0 };

/**
 * Provides the measurements for a react element at the point of layout.
 *
 * @remarks
 *
 * Taken from https://codesandbox.io/embed/lp80n9z7v9
 *
 * ```tsx
 * const [bindRef, { height }] = useMeasure()
 *
 * return <div {...bindRef}>Height: {height}</div>
 * ```
 */
export function useMeasure<Ref extends HTMLElement = any>() {
  const ref = useRef<Ref>(null);
  const [bounds, setBounds] = useState<DOMRectReadOnlyLike>(defaultBounds);

  useLayoutEffect(() => {
    const resizeObserver = new ResizeObserver(([entry]) => setBounds(entry.contentRect));

    if (ref.current) {
      resizeObserver.observe(ref.current);
    }

    return () => resizeObserver.disconnect();
  }, []);

  return [{ ref }, bounds] as const;
}

/**
 * The `setState` type for the `useSetState` hook.
 */
export type PartialDispatch<Type extends object> = (
  patch: Partial<Type> | ((prevState: Type) => Partial<Type>),
) => void;

/**
 * A `useEffect` function with a warning the provided dependencies are deeply
 * equal.
 */
const useEffectWithWarning: typeof useEffect =
  process.env.NODE_ENV === 'production'
    ? useEffect
    : (effect, deps) => {
        const ref = useRef<DependencyList>();
        const unnecessaryChange = useRef(0);

        useEffectOnce(() => {
          ref.current = deps;
        });

        useUpdateEffect(() => {
          if (!isEqual(deps, ref.current)) {
            unnecessaryChange.current = 0;
            ref.current = deps;
            return;
          }

          unnecessaryChange.current += 1;
        });

        const wrappedEffect = () => {
          if (unnecessaryChange.current >= 1) {
            console.warn(
              `The dependencies passed into your useEffect are deeply equal, but an update has been triggered ${unnecessaryChange.current} time(s). Please consider \`useMemo\` to memoize your dependencies to prevent unwanted re-renders.`,
              deps,
            );
          }

          return effect();
        };

        // eslint-disable-next-line react-hooks/exhaustive-deps
        useEffect(wrappedEffect, deps);
      };

export { useEffectWithWarning, useSetState };
