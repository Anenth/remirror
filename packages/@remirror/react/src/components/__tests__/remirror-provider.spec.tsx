import React, { FC } from 'react';

import { docNodeBasicJSON } from '@remirror/testing';
import { createReactManager, render } from '@remirror/testing/react';

import { useRemirror } from '../../hooks';
import { RemirrorProvider } from '../providers';

test('`RemirrorProvider`', () => {
  const TestComponent: FC = () => {
    const { getRootProps } = useRemirror();
    return (
      <div>
        <div data-testid='target' {...getRootProps()} />
      </div>
    );
  };

  const manager = createReactManager([]);

  const { getByRole, getByTestId } = render(
    <RemirrorProvider initialContent={docNodeBasicJSON} manager={manager}>
      <TestComponent />
    </RemirrorProvider>,
  );
  const target = getByTestId('target');
  const editor = getByRole('textbox');

  expect(target).toContainElement(editor);
});

test('multiple `getRootProps` applied to dom throw an error', () => {
  const spy = jest.spyOn(console, 'error').mockImplementation(() => {});
  const manager = createReactManager([]);
  const TestComponent: FC = () => {
    const { getRootProps } = useRemirror();

    return (
      <div>
        <div data-testid='target' {...getRootProps()} />
        <div data-testid='target1' {...getRootProps()} />
      </div>
    );
  };

  expect(() =>
    render(
      <RemirrorProvider manager={manager}>
        <TestComponent />
      </RemirrorProvider>,
    ),
  ).toThrowErrorMatchingSnapshot();

  spy.mockRestore();
});
