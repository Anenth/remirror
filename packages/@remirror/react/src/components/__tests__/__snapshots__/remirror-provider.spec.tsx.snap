// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`multiple \`getRootProps\` applied to dom throw an error 1`] = `
"\`getRootProps\` has been attached to the DOM more than once. It should only be attached to the dom once per editor.

Called 2 times

For more information visit https://remirror.io/docs/errors#rmr0201"
`;
